### ISSUE TITLE
Dart VM: Wrong full snapshot version for 1.20.4-stable

### ISSUE THREAD

**Issue by kbokarius** [acoutts/flutter-engines-no-metal#2]
@acoutts it looks like we're actually getting this error when running `flutter run --profile --cache-sksl` using the 1.20.4-stable framework files.

```VM: Wrong full snapshot version, expected '96c68176b559725593d198f1045d6b24' found '04645b6182fad3d68350d84669869ce5'```

To set our Flutter version to 1.20.4, we did a `git checkout tags/1.20.4`

Is there something else we need to do in order to ensure that those hashes match?

**Comment by kbokarius** [comment]
We did see that @rhydiant mentioned running into the same issue when building the engine himself, and he mentioned the link below, but it's not clear on how that would help us with these precompiled frameworks.

https://github.com/flutter/flutter/blob/master/bin/internal/README.md

**Comment by acoutts** [comment]
Thanks for trying! I will give it another go building that engine and try to have it uploaded tomorrow.

**Comment by kbokarius** [comment]
> Thanks for trying! I will give it another go building that engine and try to have it uploaded tomorrow.

I thought maybe I was doing something wrong or checking out the wrong Flutter branch, but it sounds like maybe something went wrong in the compilation process. I'll try again as soon as you update it, thank you!

**Comment by acoutts** [comment]
It seems a bit unclear which revision we need because the 1.20.4 versioning string isn't explicit on which revision unlike the 1.22+ versions, so perhaps I have grabbed the wrong engine revision.

Example that's easy to match:
* Flutter 1.22.0-12.0.pre - https://github.com/flutter/flutter/tree/1.22.0-12.0.pre
* Engine 1.22-candidate.12 - https://github.com/flutter/engine/commits/flutter-1.22-candidate.12

Here's how they've named it for the current stable one:
* Flutter 1.20.4
* Then engine has many 1.20 candidates so it's unclear which one we need:
<img width="448" alt="Screen Shot 2020-09-27 at 4 04 46 PM" src="https://user-images.githubusercontent.com/9597207/94374566-33184b00-00db-11eb-8b56-4c9d0f996e1d.png">

I'm going to try 1.20 revision 7 and 8 and hopefully we can find a winning combo.

**Comment by acoutts** [comment]
Alright based on your link, I think what we need is 1.20-candidate.7 because the latest commit on that branch is `d1bc06f032f9d6c148ea6b96b48261d6f545004f` and the `engine.version` file on 1.20.4 contains the commit hash `d1bc06f032f9d6c148ea6b96b48261d6f545004f`.

**Comment by kbokarius** [comment]
> Alright based on your link, I think what we need is 1.20-candidate.7 because the latest commit on that branch is `d1bc06f032f9d6c148ea6b96b48261d6f545004f` and the `engine.version` file on 1.20.4 contains the commit hash `d1bc06f032f9d6c148ea6b96b48261d6f545004f`.

Does that mean I should checkout 1.20-candidate.7 and try to use the existing frameworks you uploaded?

**Comment by acoutts** [comment]
Let me rebuild it right now on candidate 7 so you can use the latest flutter stable version, ensuring you have the latest bug fixes in place. I know there were some recent things cherry picked in for iOS 14 so you definitely want to stay on the latest stable.

**Comment by acoutts** [comment]
To clarify- I mean candidate 7 appears to be what the current flutter version is when you checkout the stable channel for 1.20.4.

**Comment by acoutts** [comment]
https://github.com/bottlepay/engine/commit/c1f28c6bf9fdfa7c2823592a8a27344d04239e84 Alright metal is disabled, doing gclient sync and building now.

**Comment by kbokarius** [comment]
@acoutts I tried checking out `flutter-1.20-candidate.7` and doing `flutter run --profile --cache-sksl` and I'm still getting the same `VM: Wrong full snapshot version` error, but maybe that wasn't supposed to fix the error?

After you upload the new `flutter-1.20-candidate.7` frameworks, should I be on the `flutter-1.20-candidate.7` branch in Flutter or a different branch/tag when building with those frameworks?

**Comment by acoutts** [comment]
Sorry, let me clarify. The thing I screwed up here is I did not compile the right engine revision for the flutter version 1.20.4. Thanks to your link in the second comment above I now understand how to properly match them.

You should checkout the flutter version tag matching the folder name in this repository, so you want to do this in your flutter folder:

```
git checkout tags/1.20.4
```

**Comment by kbokarius** [comment]
@acoutts got it, that's what I had done before, but ran into the error. It sounds like I just need to wait for you to upload the new frameworks that match the engine version. Our jank is intolerable on startup, so I'm really hoping this will alleviate it for our launch next month while not introducing any other showstopper type of issues.

**Comment by acoutts** [comment]
Sounds good! I just updated the folder names here to exactly match the flutter tag, and updated the instructions accordingly to make it more clear.

**Comment by acoutts** [comment]
Just kicked off the build. I estimate it will take my laptop about 2h to do it and I'll update you as soon as it's uploaded.

**Comment by acoutts** [comment]
Alright go ahead and try the new zip files now. Fingers crossed!

**Comment by kbokarius** [comment]
@acoutts unfortunately still getting this error:

```
2020-09-27 16:55:38.110452-0700 Runner[606:68884] [VERBOSE-3:dart_vm.cc(420)] 
Error while initializing the Dart VM: Wrong full snapshot version, expected
'8b2ca977d1d2920b9839d1b60eade6a7' found '04645b6182fad3d68350d84669869ce5'
(lldb) 
```

Edit: although the hash has now changed.

**Comment by kbokarius** [comment]
Confirming that I've run:

```
git checkout tags/1.20.4
flutter upgrade
flutter precache
```

I'm using the new profile framework and attempting to run in profile mode.

**Comment by acoutts** [comment]
Very strange - i've just tested and I can run an app it. Walk me through every step you've done from the beginning.

**Comment by acoutts** [comment]
Ahh no you're right, something still off.

**Comment by kbokarius** [comment]
I ran the following commands:

```
git checkout tags/1.20.4
flutter upgrade
flutter precache
flutter clean
```

I then downloaded ios-profile.zip from the 1.20.4 folder, unzipped it, and overwrote the files in `bin/cache/artifacts/engine/ios-profile/Flutter.framework`

Then ran `flutter run --profile --cache-sksl` which caused the app to crash each time it started, but I didn't see the error message printed. I tried running the app through XCode in profile mode and that's where I saw the error.

**Comment by acoutts** [comment]
I'll try revision 8 now.

**Comment by kbokarius** [comment]
I'm testing the  1.21.0-7.0.pre framework available here: https://github.com/flutter/flutter/issues/61045

On first try it looks like profile mode is running on the device. I'll go ahead and give it a test run and see if I can get it working and determine how much jank is alleviated and check for other issues. Although we'll definitely want to switch to the 1.20.4 stable version as soon as that's working.

**Comment by acoutts** [comment]
Sounds great. Hopefully we can find the matching combination of flutter sdk/engine for the stable channel.

**Comment by kbokarius** [comment]
Using 1.21.0-7.0.pre, as I was going through the app and starting animations to generate the SKSL file, the app randomly crashes. I doubt it's related to the version of Flutter since the crash isn't reported to Sentry, which we've only been able to reproduce in the past on an iPhone 6 Plus running iOS 12.4. The folks at Sentry said that the only type of crash that wouldn't send a crash log is an OOM crash, so my initial guess is that the app is simply using more memory with OpenGL and is crashing a lot more frequently. 

I did run the app in profile mode to check memory last week with Metal, and it was generally around 300m and would jump up to 700m when switching between screens, and it didn't seem like there was a memory leak causing the memory to constantly grow. I suppose I'll try to run the app now in profile mode with OpenGL to see how the numbers change.

**Comment by acoutts** [comment]
Good insight. I'm really confused now why the engine branch 1.20-candidate.7 didn't work for us. In flutter sdk 1.20.4 `bin/internal/engine.version` it has hash `d1bc06f032f9d6c148ea6b96b48261d6f545004f`. The latest commit in engine branch 1.20-candidate.7 is `d1bc06f032f9d6c148ea6b96b48261d6f545004f`. Seems like that should be a matching combo!

**Comment by kbokarius** [comment]
Interestingly, when I archived the release build that included the SKSL file and ran it, the app is no longer crashing. Does an app built and ran in profile mode use more memory or is it more likely to crash than an app running in release mode?

**Comment by kbokarius** [comment]
Oh, and the startup jank was definitely alleviated by packaging in the SKSL file. So far no other visual issues or other issues found other than the crashes in profile mode.

**Comment by acoutts** [comment]
That's great to hear! I think profile is pretty close to release but there is still some extra stuff included in a profile build for performance tracing.

**Comment by kbokarius** [comment]
Actually we just had the crash happen, it just seems to take longer before it happens.

**Comment by acoutts** [comment]
Hm. It could be because that's an older version of flutter too- i'd try the dev branch here and see if that one works (1.23.0)

**Comment by acoutts** [comment]
1.21 isn't on any active channel at the moment. Stable is 1.20, beta is 1.22 and dev/master are 1.23.

**Comment by kbokarius** [comment]
Ok, let me try it now.

**Comment by acoutts** [comment]
Ah shit I think I figured out the issue. I had to merge upstream 1.20-candidate.7 into my engine fork, it was way outdated! That's gotta be it. Ok, building again. Will have it uploaded tomorrow morning first thing for stable.

**Comment by kbokarius** [comment]
Ok great, I'm getting some Flutter compilation errors for 1.23 I'll try to work through tonight to see if I can do another test before the morning.

**Comment by acoutts** [comment]
For 1.22+, make sure you bump the minimum deployment target to 9.0 if you haven't yet. Also change the setting `MinimumOSVersion` in your project `ios/Flutter/App.framework/Info.plist` to `<string>9.0</string>`.

**Comment by kbokarius** [comment]
My deployment target is 12.0. I'm actually running into this issue: https://stackoverflow.com/questions/63962239/flutter-no-named-parameter-with-the-name-overflow-overflow-overflow-visible

**Comment by kbokarius** [comment]
I fixed the issues by temporarily fixing the errors in the .pub-cache folder for the dependencies. Moving ahead with the test.

**Comment by kbokarius** [comment]
Same crash with 1.23 unfortunately.

**Comment by kbokarius** [comment]
Confirmed that it's a memory issue, this is shown in XCode: Message from debugger: Terminated due to memory issue

**Comment by kbokarius** [comment]
This is actually helpful, because I didn't have the iPhone 6 Plus that this crash was reproduced on with Metal, so now I can try to debug and get to the bottom of the memory issues by running OpenGL.

**Comment by kbokarius** [comment]
After spending the last few hours trying to debug the memory crash, it's related to fade/slide animation routing with the navigator. It seems like it's inherent to Flutter, rather than something specific in our code. Interestingly searching around for memory related crashes, the iPhone 6+ is mentioned a lot which is the only device that we could reproduce the crash on with Metal, but with OpenGL we're reproducing it with an iPhone 11 Pro. At this point I'm determined to work around the crash somehow and try to get a stable build with the SKSL file using the 1.20.4-stable frameworks after you upload them tomorrow, and then we'll do a testing cycle across many different iOS devices to see if it'll be viable for our launch next month.

**Comment by acoutts** [comment]
Good digging. Sounds interesting. What sort of animations are you doing? I've not yet noticed any bugs like that on my app. Maybe there's room to improve the animation implementations on the dart side to use less resources in the meantime.

I just updated 1.20.4 and it appears to work now.

**Comment by kbokarius** [comment]
@acoutts The animation that seems most likely to crash the app is a fade transition when a new screen is pushed to the navigator. It seems to be happening with a particular screen, so there might be other aspects of that screen that are causing the issue. It's hard to pinpoint. I'll try the 1.20.4 framework shortly, thank you!

**Comment by kbokarius** [comment]
@acoutts update: we were able to mitigate the incidence of that memory crash by simplifying some of our animations. With the iPhone 6+ that we had reproduced the crash on with Metal, it seems that it's even harder to reproduce the memory crash with the OpenGL framework and modified animations. Startup jank is virtually non-existent with the new build too. At this point it looks promising, and we're going to have our QA team do a round of testing on the app tonight across several different devices and iOS versions, I'll report back but hopefully we'll be able to move ahead with this workaround and avoid having such terrible startup jank for our launch next month. I can't thank you enough for your help!
